### PR TITLE
add a lock to the messaging events api

### PR DESCRIPTION
**Pull request by calellowitz**
## Technical Summary

This is a temporary stopgap due to performance problems with this API. It should be rolled back once we have a more permanent solution. https://dimagi-dev.atlassian.net/browse/SAAS-14102

I will leave it to @dannyroberts / @snopoke to decide whether this is necessary or we can just have the partner change their plan in the meantime.

### Safety story

no changes to behavior

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by dannyroberts — APPROVED** [review]

**Comment by dannyroberts** [comment]
I jumped the gun a bit, so that approval is in theory, for once this test is fixed. The test error appears to be

```
  TypeError: acquire() got an unexpected keyword argument 'timeout'
``

File: corehq/apps/api/resources/messaging_event/view.py [hunk 1]
```diff
@@ -26,7 +28,14 @@ def messaging_events(request, domain, event_id=None):
     """Despite it's name this API is backed by the MessagingSubEvent model
     which has a more direct relationship with who the messages are being sent to.
     Each event may have more than one actual message associated with it.
+
+    Due to performance issues, this API only allows one usage at a time
     """
+
+    # if the request does not acquire the lock in one minute, return a rate limiting response
+    messaging_events_lock = get_redis_lock('MESSAGING_EVENTS_API_LOCK', timeout=60*10, name='api')
+    if not messaging_events_lock.acquire(blocking=True, timeout=60):
```
Comment (dannyroberts) [inline corehq/apps/api/resources/messaging_event/view.py:37]: I dug for a couple minutes, and I believe the parameter is called `blocking_timeout`. (Incidentally it seems like you can also specify it in the `get_redis_lock` call to set a default value, but I do like it better in the `acquire` call.)

Comment (millerdev) [inline corehq/apps/api/resources/messaging_event/view.py:37]: If this `acquire` is successful the lock will need to be released, right? I don't see where that is done.

Comment (calellowitz) [reply corehq/apps/api/resources/messaging_event/view.py:37]: :facepalm: This is a very good point

Comment (calellowitz) [reply corehq/apps/api/resources/messaging_event/view.py:37]: Huh, django redis, uses the same interface as `threading.Lock`, which uses `timeout`, but it looks like this lock is directly from the redis python library. That is my bad.

File: corehq/apps/api/resources/messaging_event/view.py [hunk 2]
```diff
@@ -34,16 +34,18 @@ def messaging_events(request, domain, event_id=None):
...
+            res = _get_individual(request, event_id)
         if request.method == 'GET' and not event_id:
-            return _get_list(request)
-        return JsonResponse({'error': "Request method not allowed"}, status=405)
+            res = _get_list(request)
+        res = JsonResponse({'error': "Request method not allowed"}, status=405)
     except BadRequest as e:
-        return JsonResponse({'error': str(e)}, status=400)
+        res = JsonResponse({'error': str(e)}, status=400)
+    messaging_events_lock.release()
```
Comment (snopoke) [inline corehq/apps/api/resources/messaging_event/view.py:47]: should this not be in a `finally` block?

**Pull request closed without merging**